In [1]:
import pandas as pd
import numpy as np

In [2]:
# I don’t think we need this cell anymore, also maybe we need records with dates before 2017 or after 2018.
# It depends on the time period we want to examine pre- post the protest (i.e 6 months or a year)  

def select_time_frame(df, date):
    df.date = pd.to_datetime(df.date)
    df = df[df['date'] > pd.to_datetime(date)]
    return df

df_la = select_time_frame(df_la, '2017-01-01')
df_pa = select_time_frame(df_pa, '2017-01-01')
df_sa = select_time_frame(df_sa, '2017-01-01')

df_la.to_csv('data/states/ca_los_angeles_2020_04_01.csv.gz', compression='gzip')
df_pa.to_csv('data/states/pa_philadelphia_2020_04_01.csv.gz', compression='gzip')
df_sa.to_csv('data/states/tx_san_antonio_2020_04_01.csv.gz', compression='gzip')



'''df_la_day = df_la.groupby(['date', 'subject_race', 'subject_sex']).size().reset_index(name='counts')
df_pa_day = df_pa.groupby(['date', 'subject_race', 'subject_sex']).size().reset_index(name='counts')
df_sa_day = df_sa.groupby(['date', 'subject_race', 'subject_sex']).size().reset_index(name='counts')

print(df_pa)

df_census = df_census
df_census_pa = df_census[df_census['state']=='Pennsylvania']

df_census_pa.head(20)
'''

### Data Wrangling

#### 1) Load, parse, filter and sort Count Love data set based on the number of Attendees 

In [3]:
df_protests = pd.read_csv('data/protests.csv')
df_protests.head(2)

Date                           Location  Attendees         Event  \
0  2017-01-15  Bowie State University, Bowie, MD     1500.0    Healthcare   
1  2017-01-16                   Johnson City, TN      300.0  Civil Rights   

                                                Tags Curated  \
0                Healthcare; For Affordable Care Act     Yes   
1  Civil Rights; For racial justice; Martin Luthe...     Yes   

                                              Source  Total Articles  
0  http://www.capitalgazette.com/news/ph-ac-cn-ac...               1  
1  http://www.johnsoncitypress.com/Local/2017/01/...               4

In [4]:
# Filter racial injustice protests
df_protests = df_protests[df_protests['Event'].str.contains('Racial Injustice')]

# Convert Date column to datetime
df_protests['Date'] = pd.to_datetime(df_protests['Date'])

# Select protests between 2017-01-01 and 2019-01-01
df_protests = df_protests[(df_protests['Date'] >= pd.to_datetime('2017-01-01')) &
                         (df_protests['Date'] < pd.to_datetime('2019-01-01'))]

# Sort protests based on attendess
df_protests = df_protests.sort_values('Attendees',ascending=False)
df_protests.head(20)

Date                          Location  Attendees  \
6875  2018-01-15                   San Antonio, TX   300000.0   
4714  2017-08-19                        Boston, MA    40000.0   
6866  2018-01-15                    Sacramento, CA    20000.0   
10539 2018-04-04                       Memphis, TN     8000.0   
6873  2018-01-15                       Phoenix, AZ     4000.0   
4907  2017-08-27                      Berkeley, CA     4000.0   
15382 2018-11-14  Drake University, Des Moines, IA     3500.0   
6897  2018-01-15                   Albuquerque, NM     3000.0   
4929  2017-08-28                    Washington, DC     3000.0   
4885  2017-08-26                     Knoxville, TN     2800.0   
5925  2017-10-19                   Gainesville, FL     2500.0   
4721  2017-08-19                        Dallas, TX     2300.0   
4720  2017-08-19                       Atlanta, GA     2000.0   
5330  2017-09-15                     St. Louis, MO     2000.0   
6900  2018-01-15                     Asheville, NC     2000.0   
5350  2017-09-16           Downtown, St. Louis, MO     2000.0   
4634  2017-08-16                  Philadelphia, PA     2000.0   
4353  2017-08-13                   Los Angeles, CA     1500.0   
4566  2017-08-14                        Eugene, OR     1500.0   
3598  2017-06-16                      St. Paul, MN     1500.0   

                                                   Event  \
6875           Racial Injustice (Martin Luther King Jr.)   
4714   Racial Injustice (Counter Protest; Alt-Right/N...   
6866           Racial Injustice (Martin Luther King Jr.)   
10539          Racial Injustice (Martin Luther King Jr.)   
6873           Racial Injustice (Martin Luther King Jr.)   
4907   Racial Injustice (Counter Protest; Alt-Right/N...   
15382                                   Racial Injustice   
6897           Racial Injustice (Martin Luther King Jr.)   
4929                                    Racial Injustice   
4885   Racial Injustice (Counter Protest; Alt-Right/N...   
5925   Racial Injustice (Counter Protest; Alt-Right/N...   
4721                  Racial Injustice (Charlottesville)   
4720                  Racial Injustice (Charlottesville)   
5330            Racial Injustice (Police Accountability)   
6900           Racial Injustice (Martin Luther King Jr.)   
5350            Racial Injustice (Police Accountability)   
4634                  Racial Injustice (Charlottesville)   
4353                  Racial Injustice (Charlottesville)   
4566                  Racial Injustice (Charlottesville)   
3598            Racial Injustice (Police Accountability)   

                                                    Tags Curated  \
6875   Civil Rights; For racial justice; Martin Luthe...     Yes   
4714   Civil Rights; For racial justice; Against whit...     Yes   
6866   Civil Rights; For racial justice; Martin Luthe...     Yes   
10539  Civil Rights; For racial justice; Martin Luthe...     Yes   
6873   Civil Rights; For racial justice; Martin Luthe...     Yes   
4907   Civil Rights; For racial justice; Against whit...     Yes   
15382                   Civil Rights; For racial justice     Yes   
6897   Civil Rights; For racial justice; Martin Luthe...     Yes   
4929                    Civil Rights; For racial justice     Yes   
4885   Civil Rights; For racial justice; Against whit...     Yes   
5925   Civil Rights; For racial justice; Against whit...     Yes   
4721   Civil Rights; For racial justice; Against whit...     Yes   
4720   Civil Rights; For racial justice; Against whit...     Yes   
5330   Civil Rights; For racial justice; For greater ...     Yes   
6900   Civil Rights; For racial justice; Martin Luthe...     Yes   
5350   Civil Rights; For racial justice; For greater ...     Yes   
4634   Civil Rights; For racial justice; Against whit...     Yes   
4353   Civil Rights; For racial justice; Against whit...     Yes   
4566   Civil Rights; For racial justice; Against whit...     Yes   
3598   Civil Rights; For 

#### 2) Select top 3 largest racial injustice protests in USA

Now that we computed the largest racial injustice protests based on the Count Love dataset, we use the summary tables provided in the website of the original paper (https://openpolicing.stanford.edu/data/) in order to find traffic stop data that matches the time period of the largest protests. Based on this table we found available data for the following protests:

1. index=6875, Date=2018-01-15, Location=San Antonio-TX, Attendees=300000
2. index=4634, Date=2017-08-16, Location=Philadelphia-PA, Attendees=2000
3. index=4353, Date=2017-08-13, Location=Los Angeles-CA, Attendees=1500

#### 3) Merge traffic stops and census data set for the 3 largest protests

##### ------------------->CENSUS FROM PAPER ...-2015

In [5]:
df_census = pd.read_csv('data/census-clean.csv.gz', compression='gzip')
df_census.head()

state  fips  year    age      race gender  count
0  Alabama  1001  2010  16-19     Asian      F   11.0
1  Alabama  1001  2010  16-19     Asian      M   18.0
2  Alabama  1001  2010  16-19     Black      F  330.0
3  Alabama  1001  2010  16-19     Black      M  386.0
4  Alabama  1001  2010  16-19  Hispanic      F   39.0

In [6]:
df_census['year'].unique()

array([2010, 2011, 2012, 2013, 2014, 2015], dtype=int64)

In [7]:
# Select census data for states of top 3 largest protests
df_census = df_census[(df_census['state'] == 'Texas') | 
                      (df_census['state'] == 'Pennsylvania') | 
                      (df_census['state'] == 'California')]

df_census = df_census[(df_census['year'] >= 2014) & (df_census['year'] <= 2015)]


df_census = df_census.groupby(['state','year','race'])['count'].sum().reset_index()
df_census.head(5)

state  year      race       count
0  California  2014     Asian   4463701.0
1  California  2014     Black   1790480.0
2  California  2014  Hispanic  10751266.0
3  California  2014     Other    885066.0
4  California  2014     White  12792449.0

##### -------------------> I USED THIS ONE for the following steps: CENSUS FROM official census.gov

see the last one-> " Age, Sex, Race, and Hispanic Origin - 5 race groups (5 race alone or in combination groups) (SC-EST2019-ALLDATA5)" https://www.census.gov/data/tables/time-series/demo/popest/2010s-state-detail.html 

In [8]:
fields = ['NAME', 'ORIGIN', 'RACE', 'AGE', 'POPESTIMATE2016', 'POPESTIMATE2017', 'POPESTIMATE2018', 'POPESTIMATE2019']
df_census = pd.read_csv('data/census2010-2019.csv', usecols=fields)
df_census.head()

NAME  ORIGIN  RACE  AGE  POPESTIMATE2016  POPESTIMATE2017  \
0  Alabama       0     1    0            37629            37526   
1  Alabama       0     1    1            37732            37904   
2  Alabama       0     1    2            38048            37975   
3  Alabama       0     1    3            37266            38170   
4  Alabama       0     1    4            37954            37322   

   POPESTIMATE2018  POPESTIMATE2019  
0            36370            35594  
1            37641            36679  
2            38172            37904  
3            38197            38510  
4            38312            38476

In [9]:
# Select census data for states of top 3 largest protests
df_census = df_census[(df_census['NAME'] == 'Texas') | 
                      (df_census['NAME'] == 'Pennsylvania') | 
                      (df_census['NAME'] == 'California')]

# Select census data for people older than 16
df_census = df_census[df_census['AGE'] >= 16]

# Exclude people with Hispanic origin and keep only Black and White races
# [1 = Not Hispanic]
df_census = df_census[df_census['ORIGIN'] == 1]

# [1 = White Alone, 2 = Black or African American Alone]
df_census = df_census[(df_census['RACE'] == 1) | (df_census['RACE'] == 2)].drop(columns='ORIGIN')

# Aggregate data based on 'NAME','ORIGIN' and 'RACE'
aggregated_columns = "POPESTIMATE2016", "POPESTIMATE2017", "POPESTIMATE2018", "POPESTIMATE2019"
df_census = df_census.groupby(['NAME', 'RACE'])[aggregated_columns].sum().reset_index()
 
df_census = df_census.rename(columns={"NAME": "state", "RACE": "race"})

df_census['race'] = df_census['race'].astype(str)

df_census['race'] [df_census['race']  == '1'] = 'white'
df_census['race'] [df_census['race']  == '2'] = 'black'

df_census.head(6)

<ipython-input-9-d7129fecc765>:18: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  df_census = df_census.groupby(['NAME', 'RACE'])[aggregated_columns].sum().reset_index()
<ipython-input-9-d7129fecc765>:24: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_census['race'] [df_census['race']  == '1'] = 'white'
<ipython-input-9-d7129fecc765>:25: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_census['race'] [df_census['race']  == '2'] = 'black'


state   race  POPESTIMATE2016  POPESTIMATE2017  POPESTIMATE2018  \
0    California  white         25416948         25265466         25065272   
1    California  black          3604796          3624044          3636574   
2  Pennsylvania  white         16518254         16452188         16387274   
3  Pennsylvania  black          2140470          2152430          2165866   
4         Texas  white         19588284         19658264         19711588   
5         Texas  black          5102892          5224160          5334814   

   POPESTIMATE2019  
0         24812928  
1          3646882  
2         16327590  
3          2178842  
4         19801104  
5          5451640

#### Stanford data from 3 largest protests  

In [10]:
df_sa = pd.read_csv('data/states/tx_san_antonio_2020_04_01.csv.gz', compression='gzip')
df_pa = pd.read_csv('data/states/pa_philadelphia_2020_04_01.csv.gz', compression='gzip')
df_la = pd.read_csv('data/states/ca_los_angeles_2020_04_01.csv.gz', compression='gzip')

C:\ProgramData\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3071: DtypeWarning: Columns (18,20,31,32,33) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,
C:\ProgramData\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3071: DtypeWarning: Columns (9) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,


In [11]:
df_sa.columns, df_pa.columns, df_la.columns

(Index(['Unnamed: 0', 'Unnamed: 0.1', 'raw_row_number', 'date', 'time',
        'location', 'lat', 'lng', 'district', 'substation', 'subject_age',
        'subject_race', 'subject_sex', 'type', 'violation', 'arrest_made',
        'citation_issued', 'outcome', 'contraband_found', 'search_conducted',
        'search_basis', 'speed', 'posted_speed', 'vehicle_color',
        'vehicle_make', 'vehicle_model', 'vehicle_registration_state',
        'vehicle_year', 'raw_race', 'raw_posted_speed', 'raw_actual_speed',
        'raw_search_reason', 'raw_contraband_or_evidence',
        'raw_custodial_arrest_made'],
       dtype='object'),
 Index(['Unnamed: 0', 'Unnamed: 0.1', 'raw_row_number', 'date', 'time',
        'location', 'lat', 'lng', 'district', 'service_area', 'subject_age',
        'subject_race', 'subject_sex', 'type', 'arrest_made', 'outcome',
        'contraband_found', 'frisk_performed', 'search_conducted',
        'search_person', 'search_vehicle', 'raw_race',
        'raw_individua

In [12]:
df_sa.subject_race.unique(), df_pa.subject_race.unique(), df_la.subject_race.unique()
# df_sa.subject_age.describe(), df_pa.subject_age.describe()
# df_sa.shape, df_pa.shape, df_la.shape

(array(['unknown', 'other', 'hispanic', 'white', 'black',
        'asian/pacific islander', nan], dtype=object),
 array(['black', 'white', 'hispanic', 'asian/pacific islander', 'unknown',
        'other'], dtype=object),
 array(['other', 'black', 'hispanic', 'white', 'asian/pacific islander'],
       dtype=object))

In [13]:
# 1.San Antonio-TX
# Aggregate stops on a daily basis and keep only black and white races
#----------------------------->>>>>> I INCLUDED PEDESTRIAN STOPS!!!!!
df_sa = df_sa[(df_sa['subject_race']=='black') | (df_sa['subject_race']=='white')]
df_sa = df_sa.groupby(['date','subject_race']).size().reset_index(name='stops')
df_sa = df_sa.rename(columns={"subject_race": "race"})
df_sa['state'] = 'Texas'

# 2. Philadelphia-PA
# Aggregate stops on a daily basis and keep only black and white races
#----------------------------->>>>>> I INCLUDED PEDESTRIAN STOPS!!!!!
df_pa = df_pa[(df_pa['subject_race']=='black') | (df_pa['subject_race']=='white')]
df_pa = df_pa.groupby(['date','subject_race']).size().reset_index(name='stops')
df_pa = df_pa.rename(columns={"subject_race": "race"})
df_pa['state'] = 'Pennsylvania'

# 3. Los Angeles-CA
# Aggregate stops on a daily basis and keep only black and white races
#----------------------------->>>>>> I INCLUDED PEDESTRIAN STOPS!!!!!
df_la = df_la[(df_la['subject_race']=='black') | (df_la['subject_race']=='white')]
df_la = df_la.groupby(['date','subject_race']).size().reset_index(name='stops')
df_la = df_la.rename(columns={"subject_race": "race"})

df_la['state'] = 'California'

In [14]:
traffic_stops_df = df_sa.append(df_pa, ignore_index=True).append(df_la, ignore_index=True)
traffic_stops_df.head()

date   race  stops  state
0  2017-01-02  black     36  Texas
1  2017-01-02  white    119  Texas
2  2017-01-03  black     46  Texas
3  2017-01-03  white    143  Texas
4  2017-01-04  black     24  Texas

In [16]:
#Merge stops with census
traffic_stops_census = pd.merge(traffic_stops_df, df_census)
traffic_stops_census.head()

date   race  stops  state  POPESTIMATE2016  POPESTIMATE2017  \
0  2017-01-02  black     36  Texas          5102892          5224160   
1  2017-01-03  black     46  Texas          5102892          5224160   
2  2017-01-04  black     24  Texas          5102892          5224160   
3  2017-01-05  black     45  Texas          5102892          5224160   
4  2017-01-06  black     35  Texas          5102892          5224160   

   POPESTIMATE2018  POPESTIMATE2019  
0          5334814          5451640  
1          5334814          5451640  
2          5334814          5451640  
3          5334814          5451640  
4          5334814          5451640